### ISSUE TITLE
Trying to use with an asset

### ISSUE THREAD

**Issue by bossanova808** [aelvan/Preparse-Field-Craft#9]
Hi Andre

Ok, so I am trying to use preparse to pre-calculate a bunch of common transforms from an image at the time of the save of that image as a new asset.  I've used similar code on entries & products but it seems to be better to do it at the asset save end of things, that way in theory everything in terms of transforms would be ready immediately for all other templates.

I'm getting errors though - my assets never finish uploading, and I get [these](url) in the craft logs:
```
Next Twig_Error_Runtime: An exception has been thrown during the rendering of a template ("HTTP status “404” encountered while attempting to download “https://images.imagescience.com.au/AledaSmallTest3.png”") in "_preparse/preparseImages" at line 14. in /var/www/vhosts/craft/craft/app/vendor/twig/twig/lib/Twig/Template.php:396
```

(See longer tack trace here - https://gist.github.com/bossanova808/626893f02a5cec2407e9)

(Might be worth adding your own plugin log to help isolate these thing by the way!)

Basically I am trying to trigger the transforms at asset save, and save a log of what happened in the field itself just so I can check what is going on.  Here's the preparse template as an example:

```
{% if asset|length %}

    {{ asset.filename }}

    {% if asset.width > 1000 %}

        Width over 1000, creating masthead.

        {% set images = craft.imager.transformImage( asset, craft.businessLogic.namedImageFit('masthead') )%}

    {% endif %}

    Creating 500 by 500 fit

    {% set images = craft.imager.transformImage( asset, craft.businessLogic.imageFit(500,500) ) %}
    
    Creating 40 by 40 thumbnail

    {% set images = craft.imager.transformImage( asset, craft.businessLogic.imageFit(40,40) ) %}


{% endif %}
```

(The craft.businessLogic just returns a transform array with my standard settings like sharpening etc, that works fine).

Any ideas why that is not working?

**Comment by aelvan** [comment]
Hi, 

Sorry for the late reply. I've tested this now, and this works for me. Did you figure out what was wrong? If not, what kind of asset source is this? And are you able to transform the images with imager from your templates?

> (Might be worth adding your own plugin log to help isolate these thing by the way!)

Imager has its own log file, but this is an exception being raised, which Craft logs to its log files.

**Comment by bossanova808** [comment]
No worries at all!  I haven't solved it, no.  The asset source is a bunch of images on S3.

Yes, I can transform them in templates just fine.  And very similar code in an entry works fine, just not on the asset.

Having a few days off for Easter so might not get back to this next week but I'll do another test then and see if I can add anything else useful here!

Have a nice break if you're having one!

**Comment by bossanova808** [comment]
Hey Andre

Ok, I just tried this again on my main dev server and it's working fine.  So must be something funky going on with my more basic home dev setup (I use LAMP at work, WAMP at home...should set up a proper LAMP or at least MAMP at home I guess!).

Sorry for the noise, this does appear to work and is *awesome* :)

**Comment by bossanova808** [comment]
Sorry andre, this has popped up again.  It's intermittent though.  The upload itself always works, but then the whirring wheel just keeps on whirring, often, while it's supposed to be doing the transforms.   

It appears to occur in sub folders, not the root.  Weirdly, when we then upload the file to the root and manually move it in to an asset subfolder, it tells us the asset exists.

For the moment I am back to letting page load do the transforms, as I'm not far from launch, but if you can think of anything let me know and I will come back to this in a few weeks once I'm past the initial launch.

**Comment by aelvan** [comment]
Ok, I'll have a look.

**Comment by aelvan** [comment]
Wow, this issue has been hanging around for ever. I've tested on several occasions, not being able to reproduce. I've also used it this way on a recent project, without experiencing any bugs. I guess you've probably moved on now, but.. Encountered it recently? 

It might be that the way Preparse now works (after the 0.3.0) is better for this kind of scenario, since it parses the fields after save. Also, there were a number of path issues in Imager that was fixed around the time you had this issue, might have had something to do with that. 

Closing this for now, let me know if you encounter this again.
